<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a data-toc-modified-id="Подключение-к-базе-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение к базе данных</a></span></li><li><span><a data-toc-modified-id="Обзор-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обзор данных</a></span><ul class="toc-item"><li><span><a data-toc-modified-id="Таблица-books-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Таблица books</a></span></li><li><span><a data-toc-modified-id="Таблица-authors-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a data-toc-modified-id="Таблица-publishers-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Таблица publishers</a></span></li><li><span><a data-toc-modified-id="Таблица-ratings-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Таблица ratings</a></span></li><li><span><a data-toc-modified-id="Таблица-reviews-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Таблица reviews</a></span></li></ul></li><li><span><a data-toc-modified-id="Задание-1-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Задание 1</a></span></li><li><span><a data-toc-modified-id="Задание-2-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Задание 2</a></span></li><li><span><a data-toc-modified-id="Задание-3-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Задание 3</a></span></li><li><span><a data-toc-modified-id="Задание-4-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Задание 4</a></span></li><li><span><a data-toc-modified-id="Задача-5-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Задача 5</a></span></li><li><span><a data-toc-modified-id="Вывод-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Вывод</a></span></li></ul></div>

# Анализ базы данных сервиса для чтения книг по подписке

**Описание проекта**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать

**Цель исследования**

Проанализировать базу данных, в которой находится информация: о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта

**Описание данных:**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора

## 1. Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
import sqlite3
from sqlalchemy import create_engine

In [2]:
# устанавка параметров
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохранение коннектора
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

## 2. Обзор данных

### 2.1 Таблица books

In [3]:
# импорт таблицы books
query_books = """

    SELECT *
    FROM books
    
"""
# вывод таблицы
books = pd.io.sql.read_sql(query_books, con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
# обзор общей информации
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [5]:
# приведён столбец publication_date к временному формату
books['publication_date'] = pd.to_datetime(books['publication_date'])
# повторный обзор общей информации
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   book_id           1000 non-null   int64         
 1   author_id         1000 non-null   int64         
 2   title             1000 non-null   object        
 3   num_pages         1000 non-null   int64         
 4   publication_date  1000 non-null   datetime64[ns]
 5   publisher_id      1000 non-null   int64         
dtypes: datetime64[ns](1), int64(4), object(1)
memory usage: 47.0+ KB


### 2.2 Таблица authors

In [6]:
# импорт таблицы authors
query_authors = """

    SELECT *
    FROM authors
    
"""
# вывод таблицы
authors = pd.io.sql.read_sql(query_authors, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
# обзор общей информации
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


### 2.3 Таблица publishers

In [8]:
# импорт таблицы publishers
query_publishers = """

    SELECT *
    FROM publishers
    
"""
# вывод таблицы
publishers = pd.io.sql.read_sql(query_publishers, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
# обзор общей информации
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


### 2.4 Таблица ratings

In [10]:
# импорт таблицы ratings
query_ratings = """

    SELECT *
    FROM ratings
    
"""
# вывод таблицы
ratings = pd.io.sql.read_sql(query_ratings, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
# обзор общей информации
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


### 2.5 Таблица reviews

In [12]:
# импорт таблицы reviews
query_reviews = """

    SELECT *
    FROM reviews
    
"""
# вывод таблицы
reviews = pd.io.sql.read_sql(query_reviews, con = engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [13]:
# обзор общей информации
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


**Вывод**

Данные никак не искажены, можно приступать к выполнению задания

## 3. Задание 1

Посчитайте, сколько книг вышло после 1 января 2000 года

In [14]:
# написание запроса
task_1 = """
    
    SELECT COUNT(DISTINCT(book_id))
    FROM books
    WHERE publication_date > '2000-01-01'
    
"""
# вывод запроса
pd.io.sql.read_sql(task_1, con = engine)

count
0    819

После 1 января 2000 года вышло **819** книг

## 4. Задание 2

Для каждой книги посчитайте количество обзоров и среднюю оценку

In [15]:
# написание запроса
task_2 = """
    
    SELECT b.book_id AS book_id,
           b.title AS title,
           COUNT(re.review_id) AS review,
           ROUND(AVG(ra.rating), 2) AS rating_avg
    FROM books as b
    LEFT JOIN reviews AS re ON b.book_id = re.book_id
    LEFT JOIN ratings AS ra ON b.book_id = ra.book_id
    GROUP BY b.book_id
    ORDER BY rating_avg DESC
    
"""
# вывод запроса
pd.io.sql.read_sql(task_2, con = engine)

book_id                                              title  review  \
0        518               Pop Goes the Weasel (Alex Cross  #5)       4   
1        732  The Ghost Map: The Story of London's Most Terr...       4   
2        347  In the Hand of the Goddess (Song of the Liones...       6   
3        610                           Tai-Pan (Asian Saga  #2)       4   
4        330  How to Be a Domestic Goddess: Baking and the A...       2   
..       ...                                                ...     ...   
995      915  The World Is Flat: A Brief History of the Twen...      12   
996      371                                              Junky       4   
997      316                  His Excellency: George Washington       4   
998      202                                      Drowning Ruth       9   
999      303                               Harvesting the Heart       4   

     rating_avg  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 4 columns]

Всех больше обзоров приходится на книгу **Twilight**, со средней оценкой 3.66

## 5. Задание 3

Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [16]:
# написание запроса
task_3 = """
    
    SELECT p.publisher_id AS publisher_id,
           p.publisher AS publisher,
           COUNT(b.book_id) AS book
    FROM books as b
    JOIN publishers AS p ON b.publisher_id  = p.publisher_id 
    WHERE num_pages > 50
    GROUP BY p.publisher_id
    ORDER BY book DESC
    LIMIT 1
    
"""
# вывод запроса
pd.io.sql.read_sql(task_3, con = engine)

publisher_id      publisher  book
0           212  Penguin Books    42

Издательство, которое выпустило наибольшее число книг толще 50 страниц является **Penguin Books**

## 6. Задание 4

Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [17]:
# написание запроса
task_4 = """
    
    SELECT request.author_id AS author_id,
           request.author AS author,
           ROUND(AVG(request.rating_avg), 2) AS rating_avg
    FROM 
        (SELECT b.book_id AS id,
                a.author_id AS author_id,
                a.author AS author,
                ROUND(AVG(r.rating), 2) AS rating_avg
         FROM books as b
         JOIN authors AS a ON b.author_id = a.author_id
         JOIN ratings AS r ON b.book_id = r.book_id
         GROUP BY b.book_id, a.author_id
         HAVING COUNT(rating_id) >= 50) AS request
         
GROUP BY author_id, author
ORDER BY rating_avg DESC  
LIMIT 1

"""
# вывод запроса
pd.io.sql.read_sql(task_4, con = engine)

author_id                      author  rating_avg
0        236  J.K. Rowling/Mary GrandPré        4.29

Автор с самой высокой оценкой, учитывая книги с 50 и более оценками является **J.K. Rowling/Mary GrandPré**

## 7. Задача 5

Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [18]:
# написание запроса
task_5 = """

SELECT ROUND(AVG(review)) AS review_avg
FROM
    (SELECT COUNT(review_id) as review
     FROM reviews
     WHERE username IN (SELECT username
                        FROM ratings
                        GROUP BY username
                        HAVING COUNT(rating_id) > 50)
     GROUP BY username) as review_username

"""
# вывод запроса
pd.io.sql.read_sql(task_5, con = engine)

review_avg
0        24.0

В среднем **24** обзора приходится от пользователей, которые поставили больше 50 оценок

## 8. Вывод

Была проделана небольшая исследовательская работа по базе данных книжного сервиса:

* После 1 января 2000 года вышло **819** книг
* Всех больше обзоров приходится на книгу **Twilight**, со средней оценкой 3.66
* Издательство, которое выпустило наибольшее число книг толще 50 страниц является **Penguin Books**
* Автор с самой высокой оценкой, учитывая книги с 50 и более оценками является **J.K. Rowling/Mary GrandPré**
* В среднем **24** обзора приходится от пользователей, которые поставили больше 50 оценок